### ISSUE TITLE
textClassification - inconsistency with  softmax temperature

### ISSUE THREAD

**Issue by carmelShor** [marcotcr/lime#416]
Hello,
First I would like to thank you for a very nice tool and for very efficient support and issues handling. 
I am trying to use LimeTextExplainer with a bert  model fine-tuned for YelpReviewsPolarity.
I was using the structure i found on https://github.com/ThilinaRajapakse/pytorch-transformers-classification .
The thing is, that when evaluating a transformer-based classifier (or any DNN for that matter), the last layer is not softmaxed, but rather argmaxed by default.  
in-order to use LimeTextExplainer, I wrote an additional method called predict_proba(str_list)--> np.array , that return the desired np array of "probabilities".
However, since softmax is a rather arbitrary way to normalize the logits into the [0 1] interval, I played a little with the softmax temperature. 
the result was a bit surprising - first of all, the "word importance" order seems to change, while i was assuming it should be invariant to monotonous transformation.
second, for some of the temperatures, the "values" each word got became very small, making it hard to display on default settings.  (see attached screenshots)
In order to check this issue, I also ran LimeTextExplainer on a default params RandomForest, with an additional softmax-temperature at the end, and got similar behavior. 

can you explain these results? 
could you elaborate on the best practices to implement the necessary predict_proba method? is there a better way to normalize (aside from softmax?)
what causes the word coefficients to vanish? 

thanks a lot,

Carmel

code:
       `

    def predict_proba(self, str_list):
        """ This function is fed to LimeTextExplainer.explain_instance() as the
        'classifier_fn'. it gets a list of strings, and gives back an np.array
        of 'probabilities', that LIME needs to fit the explanation's decision
        boundery.
        
        Arguments:
            str_list {list(str)} -- list of strings; if a single text is given,
                it should be wrapped in len=1 list.

        Returns:
            [np.array] -- an array of dim=(str_list.len, num_of_labels) of
                        the 'probabilities' for each label
        """

        dataset = self.convert_str_list_to_dataset(str_list)
        eval_sampler = SequentialSampler(dataset)
        dataloader = DataLoader(
            dataset, sampler=eval_sampler, batch_size=self.batch_size)

        preds = None
        out_label_ids = None    # might be important for other transformers

        for batch in dataloader:
            self.model.eval()
            batch = tuple(t.to(self.device) for t in batch)

            with torch.no_grad():
                inputs = {'input_ids':      batch[0],
                          'attention_mask': batch[1]
                          if self.model_type in self.transformers_models else None,
                          'token_type_ids': batch[2]
                          # XLM don't use segment_ids
                          if self.model_type in ['bert', 'xlnet'] else None}

                outputs = self.model(**inputs)
                logits = outputs[0]
                # in case model.forward gives more than one output then logits
                # should be the first
                logits = logits.detach().cpu().numpy()
                # Lime collects the logits from ALL examples, so it needs to
                # detach each batch's logits form device back to cpu.
                if preds is None:
                    preds = logits
                    # out_label_ids = inputs["labels"].detach().cpu().numpy()
                else:
                    preds = np.append(preds, logits, axis=0)
                    # out_label_ids = np.append(out_label_ids,
                    # inputs["labels"].detach().cpu().numpy(),
                    # axis=0)

        if self.softmax_T is not None:
            proba = softmaxT(preds, theta=self.softmax_T, axis=1)
            return proba
        else:
            return preds
`
screenshots:

![image](https://user-images.githubusercontent.com/58820263/71905249-71966000-3170-11ea-8fcb-96ce8f6f4ab9.png)
![image](https://user-images.githubusercontent.com/58820263/71905443-d0f47000-3170-11ea-9810-8b654ae322c9.png)
![image](https://user-images.githubusercontent.com/58820263/71905552-0c8f3a00-3171-11ea-8565-52b452f9e35c.png)

![image](https://user-images.githubusercontent.com/58820263/71906396-aacfcf80-3172-11ea-9f8f-4fd4a1cf9537.png)

**Comment by marcotcr** [comment]
>  the result was a bit surprising - first of all, the "word importance" order seems to change, while i was assuming it should be invariant to monotonous transformation.

For words that are very close together in weight, random reruns sometimes change the order. This is expected, as these are random variables. You should not put too much stock in the order if the weights are very close.

> second, for some of the temperatures, the "values" each word got became very small, making it hard to display on default settings. (see attached screenshots)

The explanation is a linear model trained on the output of the `predict_proba` function on weighted perturbed samples. If you flatten the `predict_proba` function such that perturbations don't change the prediction as much, you will tend to get get smaller weights. 

>  could you elaborate on the best practices to implement the necessary predict_proba method? is there a better way to normalize (aside from softmax?)

Ideally, you should use the predict_proba that makes sense for your problem (with whatever normalization), and use LIME to explain that. Different normalizations change what the prediction probability surface looks like, which is why you get different explanations.

Does this makes sense?

**Comment by carmelShor** [comment]
>> the result was a bit surprising - first of all, the "word importance" order seems to change, while i was assuming it should be invariant to monotonous transformation.

>For words that are very close together in weight, random reruns sometimes change the order. This is expected, as these are random variables. You should not put too much stock in the order if the weights are very close.

Well, I have set the seed to be constant (and rerun each temperature s.t. it will only use the first "draw" of the seed), so it is **not** a random artifact.

>>second, for some of the temperatures, the "values" each word got became very small, making it hard to display on default settings. (see attached screenshots)

>The explanation is a linear model trained on the output of the predict_proba function on weighted perturbed samples. If you flatten the predict_proba function such that perturbations don't change the prediction as much, you will tend to get get smaller weights.

This is my main problem - as far as I understand since the bare bert model I  was using does not care about the softmax final layer for inference (by definition it does it by argmax), and only adds softmax for the training, this choice is rather arbitrary. 
when I fed lime with the bare logits from bert model (not normalized for [0 ,1]) I got some nice looking explanations but was a bit hesitant about the interpretation. [see screenshot below]

>>could you elaborate on the best practices to implement the necessary predict_proba method? is there a better way to normalize (aside from softmax?)

>Ideally, you should use the predict_proba that makes sense for your problem (with whatever normalization), and using LIME to explain that. Different normalizations change what the prediction probability surface looks like, which is why you get different explanations.

The problem is that actually there isn't a way "that makes sense for your problem", or to the matter of fact any problem, the mismatch between probability softmax and confidence is a matter for much deeper discussion (cf. https://arxiv.org/pdf/1706.04599.pdf). 
but for now, my next stage goal is to use lime (or sp-lime) to get a "set of all important words" for each class, and later use this lists to generate free-text explanations. for that matter I would like the weights of each explanation to be consistent with all other explanations, and I am afraid they are currently not.

>Does this makes sense?
![image](https://user-images.githubusercontent.com/58820263/71994022-bab5e500-3240-11ea-8cdd-0ac177e87e0e.png)

![image](https://user-images.githubusercontent.com/58820263/71993950-9c4fe980-3240-11ea-9015-1c77a8d5205f.png)

**Comment by marcotcr** [comment]
> Well, I have set the seed to be constant (and rerun each temperature s.t. it will only use the first "draw" of the seed), so it is not a random artifact.
Hm, this is somewhat puzzling then. I'm wondering if the regularization is doing something here.

> This is my main problem - as far as I understand since the bare bert model I was using does not care about the softmax final layer for inference (by definition it does it by argmax), and only adds softmax for the training, this choice is rather arbitrary.
when I fed lime with the bare logits from bert model (not normalized for [0 ,1]) I got some nice looking explanations but was a bit hesitant about the interpretation. [see screenshot below]

Even if inference is done by argmax, you have to pick some scale in which explanations will be displayed (i.e. what is being 'moved' as you perturb different features). You certainly could use a binary representation (i.e. just the argmax) as the thing you want to explain, but you tend to lose the signal present in variations in the logits for predictions of the same class. 
Taking your bert screenshots as examples: when you get a weight of 0.93 for 'worst' with 'no softmax', that roughly means 'if you remove the word worst, in general you move the logits by 0.93', whereas with softmax it means 'if you remove it, in general you move the prediction probability by 0.07'. Those two are equivalent (minus the shrinking done by regularization, which will be different).

> The problem is that actually there isn't a way "that makes sense for your problem", or to the matter of fact any problem, the mismatch between probability softmax and confidence is a matter for much deeper discussion (cf. https://arxiv.org/pdf/1706.04599.pdf).
Indeed, a lot could be discussed here. In practice though, everyone picks something and typically thresholds based on that. Since that is what is going to control the actual decisions of the model, that is what should be used with LIME : )

>  but for now, my next stage goal is to use lime (or sp-lime) to get a "set of all important words" for each class, and later use this lists to generate free-text explanations. for that matter I would like the weights of each explanation to be consistent with all other explanations, and I am afraid they are currently not.
If this is your goal, then I would definitely use some normalization on the outputs to make them more consistent across instances.

Sorry for the long delay in responding!

**Comment by ccarmel** [comment]
thanks for your response!

in a strange turn of events, i'm working on Lime implementation for Bert models again, I have posed more issues on a separate thread.

regarding your reply, I will have a more thorough masking investigation of the importance scores I get with and without softmax and T-softmax and will post my findings here.

**Comment by marcotcr** [comment]
Sounds good : )
I'll close for now, and reopen if necessary
